src/geometry/manifold/real_instances.lean

```diff
@@ -0,0 +1,375 @@
+  to_fun := λ(z : Icc x y), ⟨λi, y - z.val, sub_nonneg.mpr z.property.2⟩,
+  inv_fun := λz, ⟨max (y - z.val 0) x, by simp [le_refl, z.property, le_of_lt h]⟩,
+  map_source := by simp,
+  map_target := by { simp, assume z hz, left, linarith },
+  left_inv := by { rintros ⟨z, hz⟩ h'z, simp at hz h'z, simp [hz, h'z] },
+  right_inv := begin
+    rintros ⟨z, hz⟩ h'z,
+    rw subtype.mk_eq_mk,
+    funext,
+    simp at hz h'z,
+    have A : x ≤ y + -z 0, by linarith,
+    rw subsingleton.elim i 0,
+    simp [A],
+  end,
+  open_source := begin
+    have : is_open {z : ℝ | x < z} := is_open_Ioi,
+    exact continuous_subtype_val _ this
+  end,
+  open_target := begin
+    have : is_open {z : ℝ | z < y - x} := is_open_Iio,
+    have : is_open {z : fin 1 → ℝ | z 0 < y - x} :=
+      (continuous_apply 0) _ this,
+    exact continuous_subtype_val _ this
+  end,
+  continuous_to_fun := begin
+    apply continuous.continuous_on,
+    apply continuous_subtype_mk,
+    have : continuous (λ (z : ℝ) (i : fin 1), y - z) :=
+      continuous_sub continuous_const (continuous_pi (λi, continuous_id)),
+    exact continuous.comp this continuous_subtype_val,
+  end,
+  continuous_inv_fun := begin
+    apply continuous.continuous_on,
+    apply continuous_subtype_mk,
+    have A : continuous (λ z : ℝ, max (y - z) x) :=
+      continuous_max (continuous_sub continuous_const continuous_id) continuous_const,
+    have B : continuous (λz : fin 1 → ℝ, z 0) := continuous_apply 0,
+    exact (A.comp B).comp continuous_subtype_val
+  end }
+
+/--
+Manifold with boundary structure on [x, y], using only two charts.
+-/
+instance Icc_manifold (x y : ℝ) [h : lt_class x y] : manifold (euclidean_half_space 1) (Icc x y) :=
+{ atlas := {Icc_left_chart x y, Icc_right_chart x y},
+  chart_at := λz, if z.val < y then Icc_left_chart x y else Icc_right_chart x y,
+  mem_chart_source := λz, begin
+    by_cases h' : z.val < y,
+    { simp only [h', if_true],
+      exact h' },
+    { simp only [h', if_false],
+      apply lt_of_lt_of_le h,
+      simpa using h' }
+  end,
+  chart_mem_atlas := λz, by { by_cases h' : z.val < y; simp [h'] } }
+
+/--
+The manifold structure on [x, y] is smooth.
+-/
+instance Icc_smooth_manifold (x y : ℝ) [lt_class x y] :
+  smooth_manifold_with_corners (model_with_corners_euclidean_half_space 1) (Icc x y) :=
+begin
+  have M : times_cont_diff_on ℝ ⊤ (λz : fin 1 → ℝ, (λi : fin 1, y - x) - z) univ,
+  { rw times_cont_diff_on_univ,
+    exact times_cont_diff.sub times_cont_diff_const times_cont_diff_id },
+  haveI : has_groupoid (Icc x y) (times_cont_diff_groupoid ⊤ (model_with_corners_euclidean_half_space 1)) :=
+  begin
+    apply has_groupoid_of_pregroupoid,
+    assume e e' he he',
+    simp [atlas] at he he',
+    /- We need to check that any composition of two charts gives a C^∞ function. Each chart can be
+    either the left chart or the right chart, leaving 4 possibilities that we handle successively.
+    -/
+    cases he; cases he',
```

Comment:
Possible proof golf:
```lean
    rcases he with rfl|rfl; rcases he' with rfl|rfl,
    { -- e = right chart, e' = right chart
      refine ((mem_groupoid_of_pregroupoid _ _).mpr _).1,
      apply symm_trans_mem_times_cont_diff_groupoid },
```
delete the hypotheses `he` and `he'` from `simp` lists, and use the same proof as above in the fourth case.

Reply:
Very nice, thanks